### PR TITLE
Support batch trigger endpoint

### PR DESCRIPTION
**Pull request by mdpye**
First pass, improvements to interface expected

### CODE REVIEW

File: PusherServer.Tests/AcceptanceTests/Trigger.cs [hunk 1]
```diff
@@ -175,6 +176,71 @@ public void It_should_return_a_200_response_async()
...
+            IPusher pusher = new Pusher(Config.AppId, Config.AppKey, Config.AppSecret, new PusherOptions()
+            {
+                HostName = Config.Host
+            });
+
+            BatchEvent[] events = new BatchEvent[] {
+                new BatchEvent {
+                    channel = "my-channel-1",
+                    name = "my_event",
+                    data = new JsonSerializer().Serialize(new { hello = "world" })
```
Comment (mdpye) [inline PusherServer.Tests/AcceptanceTests/Trigger.cs]: I don't like that I have to push serialisation back to the user here - what's the correct way to deal with this slickly. Ideally, we want to supply a dictionary when constructing the BatchEvent, but serialise it to a string and have the type of the property be `string` (for the later serialisation pass).

Can this be done together with the slick property initialisation syntax, or do I need to expose different objects to the end user and then translate in to `BatchEvent`s for serialisation?

File: PusherServer/IPusher.cs [hunk 2]
```diff
@@ -60,6 +60,8 @@ public interface IPusher
         /// For any non-200 response from the HTTP API or if the response body cannot be parsed as JSON
         /// </exception>
         ITriggerResult Trigger(string[] channelNames, string eventName, object data, ITriggerOptions options);
+
+        ITriggerResult Trigger(BatchEvent[] events);
```
Comment (mdpye) [inline PusherServer/IPusher.cs]: Is there a more friendly "list" type, or is array the standard in C#?

**Comment by mdpye** [comment]
@leggetter please review for idiomatic C# - I've left a couple of specific queries as comments already.

File: PusherServer.Tests/AcceptanceTests/Trigger.cs (hunk 1 above)
Comment (leggetter) [inline PusherServer.Tests/AcceptanceTests/Trigger.cs]: You should be able to make the `data` property an `object` from the API point of view. Not sure about how the internals work - but I'll take a look at the code in full.

File: PusherServer/IPusher.cs (hunk 2 above)
Comment (leggetter) [inline PusherServer/IPusher.cs]: Using the low-level Array is nice enough here IMO. It's easy for a developer to use a more complex data type and then convert it to an Array when calling `Trigger`.

File: PusherServer.Tests/AcceptanceTests/Trigger.cs [hunk 3]
```diff
@@ -5,6 +5,7 @@
 using NUnit.Framework;
 using System.IO;
 using System.Web.Script.Serialization;
+using RestSharp.Serializers;
```
Comment (leggetter) [inline PusherServer.Tests/AcceptanceTests/Trigger.cs]: Hopefully we can remove this.

File: PusherServer/BatchEvent.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        /// <summary>
+        /// The name of the event
+        /// </summary>
+        public string name { get; set; }
+
+        /// <summary>
+        /// The event data
+        /// </summary>
+        public string data { get; set; }
```
Comment (leggetter) [inline PusherServer/BatchEvent.cs]: I'd suggest we make this an `object` from a public API point of view. We should also consider making the properties upper-case to adhere to the C# convention.

File: PusherServer/BatchTriggerBody.cs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+﻿
+namespace PusherServer
+{
+    /// <summary>
+    /// Represents the payload to be sent when triggering events
+    /// </summary>
+    class BatchTriggerBody
+    {
+        public BatchEvent[] batch { get; set; }
```
Comment (leggetter) [inline PusherServer/BatchTriggerBody.cs]: `batch` -> `Events` ?

File: PusherServer/IPusher.cs [hunk 6]
```diff
@@ -1,4 +1,4 @@
-﻿using System;
+﻿        using System;
```
Comment (leggetter) [inline PusherServer/IPusher.cs]: Accidental tab?

File: PusherServer/BatchEvent.cs [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+{
+    /// <summary>
+    /// Represents an event that is part of a batch trigger
+    /// </summary>
+    public class BatchEvent
+    {
+        /// <summary>
+        /// The name of the event
+        /// </summary>
+        public string name { get; set; }
```
Comment (leggetter) [inline PusherServer/BatchEvent.cs]: `name` -> `Name`

File: PusherServer/BatchEvent.cs [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+        /// <summary>
+        /// The event data
+        /// </summary>
+        public string data { get; set; }
+
+        /// <summary>
+        /// The channel the event should be triggered on.
+        /// </summary>
+        public string channel { get; set; }
```
Comment (leggetter) [inline PusherServer/BatchEvent.cs]: `channel` -> `Channel`

File: PusherServer/BatchEvent.cs [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+        /// <summary>
+        /// The id of a socket to be excluded from receiving the event.
+        /// </summary>
+        public string socket_id { get; set; }
```
Comment (leggetter) [inline PusherServer/BatchEvent.cs]: `socket_id` -> `SocketId`

File: PusherServer/BatchTriggerBody.cs (hunk 5 above)
Comment (leggetter) [inline PusherServer/BatchTriggerBody.cs]: Maybe this isn't a public API so it may be fine. Although there are ways of controlling the name of the serialised property.

File: PusherServer/BatchEvent.cs (hunk 4 above)
Comment (mdpye) [inline PusherServer/BatchEvent.cs]: Both will fall foul of the DefaultJSONSerialiser: capital property names (I imagine) will become capital JSON field names, which are not valid. An object will be serialised in-line, not treated as opaque data.

Comment (mdpye) [inline PusherServer/BatchEvent.cs]: We can probably fix the property naming to match the convention by tuning the JSON serialiser (you'll see snake case property names elsewhere at the moment as well), but I think this is out of scope for this branch.

File: PusherServer/BatchTriggerBody.cs (hunk 5 above)
Comment (mdpye) [inline PusherServer/BatchTriggerBody.cs]: It has to be `batch`, it's the key we desire in the serialised output. Unless there are "annotations" we can add to tell the serialiser otherwise.

File: PusherServer.Tests/AcceptanceTests/Trigger.cs (hunk 1 above)
Comment (mdpye) [inline PusherServer.Tests/AcceptanceTests/Trigger.cs]: OK, I'll use different objects for API and serialisation.

File: PusherServer/BatchEvent.cs (hunk 4 above)
Comment (leggetter) [inline PusherServer/BatchEvent.cs]: I think a `SerializeAs` attribute may be supported. See: https://github.com/restsharp/RestSharp/blob/658f0984e3ec71f2119f50f7de878f155246e9ce/RestSharp.Tests/SerializerTests.cs#L261

We'll need to test.

File: PusherServer/BatchTriggerBody.cs (hunk 5 above)
Comment (leggetter) [inline PusherServer/BatchTriggerBody.cs]: If it's internal - I'm not fussed.

File: PusherServer/BatchEvent.cs (hunk 4 above)
Comment (leggetter) [inline PusherServer/BatchEvent.cs]: If not, we'll need to use another object.

**Comment by leggetter** [comment]
3.0.0-rc2 is now on Nuget. Install using

```
Install-Package PusherServer -Pre
